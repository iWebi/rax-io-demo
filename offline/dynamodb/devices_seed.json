[
  {
    "ip": "10.0.0.1",
    "id": "12345",
    "name": "device-1",
    "hashKey": "network_device",
    "rangeKey": "12345",
    "created": "2022-03-06",
    "updated": "2022-03-06"
  },
  {
    "ip": "10.0.0.2",
    "id": "22345",
    "name": "device-2",
    "hashKey": "network_device",
    "rangeKey": "22345",
    "created": "2022-03-06",
    "updated": "2022-03-06"
  }
]
