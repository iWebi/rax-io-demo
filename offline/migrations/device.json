{
  "Table": {
    "TableName": "Device",
    "AttributeDefinitions": [
      { "AttributeName": "hashKey", "AttributeType": "S" },
      { "AttributeName": "rangeKey", "AttributeType": "S" }
    ],
    "KeySchema": [
      { "AttributeName": "hashKey", "KeyType": "HASH" },
      { "AttributeName": "rangeKey", "KeyType": "RANGE" }
    ],
    "ProvisionedThroughput": {
      "ReadCapacityUnits": 1,
      "WriteCapacityUnits": 1
    }
  }
}
