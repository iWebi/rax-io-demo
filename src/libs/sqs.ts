import { SendMessageCommand, SendMessageCommandInput, SQSClient } from "@aws-sdk/client-sqs";
import * as AWSXRay from "aws-xray-sdk";
import { Device } from "./types";

const sqsClient = buildSQSClient();

function buildSQSClient() {
  const { IS_OFFLINE } = process.env;
  const offlineOptions = {
    region: "localhost",
    endpoint: "http://localhost:9324",
    // either .aws/credentials or ENV variables or config needs credentials keys.
    // These are checked by AWS SDK but not used in offline mode
    credentials: {
      accessKeyId: "required-but-not-used-weird",
      secretAccessKey: "required-but-not-used-weird",
    },
  };
  return IS_OFFLINE
    ? new SQSClient(offlineOptions)
    : AWSXRay.captureAWSv3Client(new SQSClient({ region: process.env.AWS_REGION }));
}

function getNotificationsQueueName(): string {
  const { IS_OFFLINE, NOTIFICATION_QUEUE_URL } = process.env;
  return IS_OFFLINE ? "http://localhost:9324/queue/devicenotifications.fifo" : NOTIFICATION_QUEUE_URL;
}

export async function sendNewDeviceMessage(device: Device) {
  const sendMessageInput = {
    QueueUrl: getNotificationsQueueName(),
    MessageBody: JSON.stringify(device),
    MessageGroupId: `Devices-${device.id}-${device.tenantId}`,
  } as SendMessageCommandInput;
  await sqsClient.send(new SendMessageCommand(sendMessageInput));
}
