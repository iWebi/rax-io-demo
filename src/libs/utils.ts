import { APIGatewayProxyResult } from "aws-lambda";
import { DEFAULT_RESPONSE_HEADERS } from "./constants";
import { AppError, DynamoItemResponse } from "./types";

export function setStatusType(code: any): string {
  switch (code) {
    case 400:
      return "Bad Request";
    case 401: // intentional UnAuthorized instead of Not Authenticated error
    case 403:
      return "UnAuthorized";
    case 404:
      return "Not Found";
    case 408:
      return "Request Timeout";
    case 500:
      return "Internal Server Error";
    default:
      return "";
  }
}

export function debug(message: string, ...optionalParams: any[]) {
  console.log(`DEBUG: ${message}`, optionalParams);
}

export function badRequestWith(message?: string): AppError {
  return {
    body: message ?? "",
    statusCode: 400,
    statusType: "Bad Request",
  };
}

export function internalServerErrorWith(message?: string): AppError {
  return {
    body: message ?? "",
    statusCode: 500,
    statusType: "Internal Server Error",
  };
}

export function errorProxyResponse(error: AppError): APIGatewayProxyResult {
  return {
    body: JSON.stringify(error),
    headers: DEFAULT_RESPONSE_HEADERS,
    statusCode: error.statusCode,
  };
}

export function badRequestProxyResponse(message?: string): APIGatewayProxyResult {
  return errorProxyResponse(badRequestWith(message));
}

export function successProxyResponse(data?: any, statusCode?: number): APIGatewayProxyResult {
  return {
    body: data ? JSON.stringify(data) : "",
    statusCode: statusCode ?? 200,
    headers: DEFAULT_RESPONSE_HEADERS,
    isBase64Encoded: false,
  };
}

export function isEmpty(value: any): boolean {
  return value === null || value === undefined;
}

export function isNonEmpty(value: any): boolean {
  return !isEmpty(value);
}

export function okResponse<T>(data: T): DynamoItemResponse<T> {
  return {
    body: data,
    statusCode: 200,
    statusType: "OK",
  };
}

export function notFoundResponse(message?: string): DynamoItemResponse<string> {
  return {
    body: message ?? "",
    statusCode: 404,
    statusType: setStatusType(404),
  };
}
