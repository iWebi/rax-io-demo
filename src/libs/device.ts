import { APIGatewayProxyResult } from "aws-lambda";
import * as repository from "./devicerepository";
import { sendNewDeviceMessage } from "./sqs";
import { Device } from "./types";
import { badRequestWith, errorProxyResponse, isEmpty, successProxyResponse } from "./utils";

export async function addDevice(deviceRequest: string, tenant: string): Promise<APIGatewayProxyResult> {
  try {
    const device = validateDeviceForCreateOrUpdate(deviceRequest);
    device.tenantId = tenant;
    await repository.addDevice(device);
    await sendNewDeviceMessage(device);
    return successProxyResponse(device, 200);
  } catch (err) {
    return errorProxyResponse(err);
  }
}

function validateDeviceForCreateOrUpdate(deviceRequest: string): Device {
  let device: Device;
  try {
    device = JSON.parse(deviceRequest);
  } catch (err) {
    throw badRequestWith("Invalid JSON payload");
  }
  // Example validations
  if (isEmpty(device.name) || isEmpty(device.ip)) {
    throw badRequestWith("Missing mandatory fields: name, ip for device creation");
  }
  return device;
}

export async function getDevice(deviceId: string, tenant: string): Promise<APIGatewayProxyResult> {
  try {
    const deviceResponse = await repository.getDevice(deviceId, tenant);
    return successProxyResponse(deviceResponse.body, 200);
  } catch (err) {
    return errorProxyResponse(err);
  }
}

export async function deleteDevice(deviceId: string, tenant: string): Promise<APIGatewayProxyResult> {
  try {
    const deleteResponse = await repository.deleteDevice(deviceId, tenant);
    return successProxyResponse(deleteResponse.body, 200);
  } catch (err) {
    return errorProxyResponse(err);
  }
}

export async function updateDevice(
  deviceRequest: string,
  deviceId: string,
  tenant: string
): Promise<APIGatewayProxyResult> {
  try {
    const device = validateDeviceForCreateOrUpdate(deviceRequest);
    device.id = deviceId;
    device.tenantId = tenant;
    const updateResponse = await repository.updateDevice(device);
    return successProxyResponse(updateResponse.body);
  } catch (err) {
    return errorProxyResponse(err);
  }
}
