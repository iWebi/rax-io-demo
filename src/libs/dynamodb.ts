import {
  AttributeValue,
  DeleteItemOutput,
  DynamoDBClient,
  QueryOutput,
  ScanOutput,
  UpdateItemOutput,
} from "@aws-sdk/client-dynamodb";
import {
  DeleteCommand,
  DynamoDBDocumentClient,
  GetCommand,
  PutCommand,
  QueryCommand,
  UpdateCommand,
} from "@aws-sdk/lib-dynamodb";
import * as AWSXRay from "aws-xray-sdk";
import { debug } from "console";
import DefaultClientConfig from "./awsclientconfig";
import { DynamoDBQueryResponse, DynamoItemResponse } from "./types";
import { internalServerErrorWith, notFoundResponse, okResponse, setStatusType } from "./utils";

const ddbDocClient = buildDocClient();

export async function getItem(params: any): Promise<DynamoItemResponse<any>> {
  const data = await ddbDocClient.send(new GetCommand(params));
  if (!data.Item) {
    throw404(params);
  }
  return okResponse(data.Item!);
}

export async function getItems(params: any): Promise<DynamoDBQueryResponse<any>> {
  const response = await ddbDocClient.send(new QueryCommand(params));
  if (response.Count! === 0) {
    throw404(params);
  }
  return collectionResponse(response);
}

export async function putItem(params: any): Promise<DynamoItemResponse<any>> {
  try {
    await ddbDocClient.send(new PutCommand(params));
  } catch (err) {
    throw errorResponse(err);
  }
  return okResponse("");
}

export async function updateItem(params: any): Promise<DynamoItemResponse<any>> {
  let data: UpdateItemOutput = {};
  try {
    data = await ddbDocClient.send(new UpdateCommand(params));
  } catch (err) {
    throw errorResponse(err);
  }
  return okResponse([data.Attributes!]);
}

export async function deleteItem(params: any): Promise<DynamoItemResponse<any>> {
  let data: DeleteItemOutput = {};
  try {
    data = await ddbDocClient.send(new DeleteCommand(params));
  } catch (err) {
    throw errorResponse(err);
  }
  if (!data.Attributes) {
    throw404(params);
  }
  return okResponse(data.Attributes!);
}

function errorResponse(err: any) {
  if (err.hasOwnProperty("$metadata")) {
    const code = err.httpStatusCode;
    return {
      body: err.message,
      statusCode: code,
      statusType: setStatusType(code),
    };
  }
  return internalServerErrorWith(JSON.stringify(err));
}

function collectionResponse(data: ScanOutput | QueryOutput) {
  const output = data.Items;
  return {
    Count: output?.length,
    LastEvaluatedKey: data.LastEvaluatedKey,
    body: output,
    statusCode: 200,
    statusType: "OK",
  };
}

export function toDynamoString(value: string): AttributeValue {
  return { S: value ?? null };
}

export function toDynamoNumber(value: number) {
  return { N: value };
}

function throw404(params: any, message?: string) {
  debug("Item not found for params", params);
  throw notFoundResponse(message);
}

export function buildDocClient() {
  const marshallOptions = {
    // Whether to automatically convert empty strings, blobs, and sets to `null`.
    convertEmptyValues: false, // false, by default.
    // Whether to remove undefined values while marshalling.
    removeUndefinedValues: true, // false, by default.
    // Whether to convert typeof object to map attribute.
    convertClassInstanceToMap: false, // false, by default.
  };

  const unmarshallOptions = {
    // Whether to return numbers as a string instead of converting them to native JavaScript numbers.
    wrapNumbers: false, // false, by default.
  };

  const { IS_OFFLINE } = process.env;
  const offlineOptions = {
    region: "localhost",
    endpoint: "http://localhost:8000",
    // either .aws/credentials or ENV variables or config needs credentials keys.
    // These are checked by AWS SDK but not used in offline mode
    credentials: {
      accessKeyId: "required-but-not-used-weird",
      secretAccessKey: "required-but-not-used-weird",
    },
  };
  const dynamodb = IS_OFFLINE
    ? new DynamoDBClient(offlineOptions)
    : AWSXRay.captureAWSv3Client(new DynamoDBClient(DefaultClientConfig));
  return DynamoDBDocumentClient.from(dynamodb, { marshallOptions, unmarshallOptions });
}
