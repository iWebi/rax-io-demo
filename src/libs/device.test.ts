import * as ulid from "ulid";
import { addDevice } from "./device";
import * as dynamodb from "./dynamodb";
import * as sqs from "./sqs";
import { okResponse } from "./utils";

jest.mock("./dynamodb");
jest.mock("./sqs");
jest.mock("ulid");

const mockedDynamoDB = dynamodb as jest.Mocked<typeof dynamodb>;
const mockedSqs = sqs as jest.Mocked<typeof sqs>;
const mockedUlid = ulid as jest.Mocked<typeof ulid>;

const mockDate = "2022-03-11T06:28:16.263Z";
jest.spyOn(Date.prototype, "toISOString").mockImplementation(() => mockDate);
mockedUlid.ulid.mockReturnValue("test-random-ulid-value");

const tenant = "test-tenant-12345";

describe("addDevice", () => {
  it("returns error for invalid json", async () => {
    // given
    const deviceRequest = "not a valid json";
    // when
    const response = await addDevice(deviceRequest, tenant);
    // then
    expect(response).toMatchSnapshot();
  });

  it("returns error for missing name field", async () => {
    // given
    const deviceRequest = '{"ip": "10.0.0.1"}';
    // when
    const response = await addDevice(deviceRequest, tenant);
    // then
    expect(response).toMatchSnapshot();
  });

  it("inserts item to DynamoDB for valid request", async () => {
    // given
    const deviceRequest = '{"ip": "10.0.0.1", "name": "test-device-1"}';
    mockedDynamoDB.putItem.mockResolvedValueOnce(okResponse(""));
    // when
    const response = await addDevice(deviceRequest, tenant);
    // then
    expect(response).toMatchSnapshot();
    expect(mockedDynamoDB.putItem.mock.calls).toMatchSnapshot("dynamodb insert payload");
    expect(mockedSqs.sendNewDeviceMessage.mock.calls).toMatchSnapshot("sqs notification payload");
  });
});
