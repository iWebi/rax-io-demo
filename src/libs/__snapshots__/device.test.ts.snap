// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addDevice inserts item to DynamoDB for valid request 1`] = `
Object {
  "body": "{\\"body\\":\\"\\",\\"statusCode\\":200,\\"statusType\\":\\"OK\\"}",
  "headers": Object {
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": 200,
}
`;

exports[`addDevice inserts item to DynamoDB for valid request 2`] = `
Array [
  Array [
    Object {
      "Item": Object {
        "created": "2022-03-11T06:28:16.263Z",
        "hashKey": "test-tenant-12345-Devices",
        "id": "test-random-ulid-value",
        "ip": "10.0.0.1",
        "name": "test-device-1",
        "rangeKey": "test-random-ulid-value",
        "tenantId": "test-tenant-12345",
        "updated": "2022-03-11T06:28:16.263Z",
      },
      "TableName": "Device",
    },
  ],
]
`;

exports[`addDevice returns error for invalid json 1`] = `
Object {
  "body": "{\\"body\\":\\"Invalid JSON payload\\",\\"statusCode\\":400,\\"statusType\\":\\"Bad Request\\"}",
  "headers": Object {
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": 400,
}
`;

exports[`addDevice returns error for missing name field 1`] = `
Object {
  "body": "{\\"body\\":\\"Missing mandatory fields: name, ip for device creation\\",\\"statusCode\\":400,\\"statusType\\":\\"Bad Request\\"}",
  "headers": Object {
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": 400,
}
`;
